lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -820,61 +818,82 @@ def _bulk_dispatch(
+                    results = helpers.streaming_bulk(
+                        elastic,
+                        self.generate_actions(dataset, context, map),
+                        raise_on_exception=False,
+                        raise_on_error=False,
+                    )
+                    report = self._analyze_bulk(results, context)
+                except Exception as e:
+                    raise APIInternalError("Unexpected backend error '{e}'") from e
+            elif context["attributes"].require_map and self.expect_index(dataset):
+                # If the dataset has no index map, the bulk operation requires one,
+                # and we expect one to appear, fail rather than risking abandoning
+                # Elasticsearch documents.
+                raise APIAbort(
+                    HTTPStatus.CONFLICT,
+                    f"Operation unavailable: dataset {dataset.resource_id} is not indexed.",
+                )
+            else:
+                report = BulkResults(errors=0, count=0, report={})
 
-        # If we don't have an Elasticsearch index map, then the dataset isn't
-        # indexed and we skip the Elasticsearch actions.
-        if IndexMap.exists(dataset):
-            # Build an Elasticsearch instance to manage the bulk update
-            elastic = Elasticsearch(self.elastic_uri)
-            map = IndexMap.stream(dataset=dataset)
+            summary: JSONOBJECT = {
+                "ok": report.count - report.errors,
+                "failure": report.errors,
+            }
+            attributes: JSONOBJECT = {"summary": summary}
+            auditing["attributes"] = attributes
 
-            # NOTE: because both generate_actions and streaming_bulk return
-            # generators, the entire sequence is inside a single try block.
+            # Let the subclass complete the operation
             try:
-                results = helpers.streaming_bulk(
-                    elastic,
-                    self.generate_actions(dataset, context, map),
-                    raise_on_exception=False,
-                    raise_on_error=False,
-                )
-                report = self._analyze_bulk(results, context)
+                self.complete(dataset, context, summary)
             except Exception as e:
-                raise APIInternalError("Unexpected backend error '{e}'") from e
-        elif context["attributes"].require_map and self.expect_index(dataset):
-            # If the dataset has no index map, the bulk operation requires one,
-            # and we expect one to appear, fail rather than risking abandoning
-            # Elasticsearch documents.
-            raise APIAbort(
-                HTTPStatus.CONFLICT,
-                f"Operation unavailable: dataset {dataset.resource_id} is not indexed.",
-            )
-        else:
-            report = BulkResults(errors=0, count=0, report={})
+                raise APIInternalError(f"Unexpected completion error '{e}'") from e
```

Comment:
I'm not sure whether any of the current `ElasticBulkBase` subclasses do, but the code was prepared to handle it gracefully and I saw no point in changing that.

Reply:
I looked, and I didn't see any, which is a good thing, because I _don't_ think the result will be graceful:  I think the code at line 873 would raise an `APIInternalError` reporting either an unexpected (nested) `APIInternalError` or `APIAbort`, rather than re-raising the original (nested) exception.  In the case of a nested `APIAbort`, that would be a problem (because it would change the result from a `4XX` to a `500`); in the case of an `APIInternalError`, it won't have much user impact but it'll be a little weird in the log.

Reply:
Again, I didn't change the logic here, and I didn't really want to worry about it. But, sure; it would be a bit cleaner to exclude `APIAbort` "just in case".